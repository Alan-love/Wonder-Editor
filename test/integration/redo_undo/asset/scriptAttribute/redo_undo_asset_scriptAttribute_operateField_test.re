open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe("redo_undo: asset script attribute->operate field", () => {
    let sandbox = getSandboxDefaultVal();

    beforeEach(() => {
      sandbox := createSandbox();
      MainEditorSceneTool.initState(~sandbox, ());
      MainEditorSceneTool.createDefaultScene(
        sandbox,
        MainEditorSceneTool.setFirstCubeToBeCurrentSceneTreeNode,
      );
    });
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));

    describe("test rename field", () => {
      let _simulate = () => {
        let assetTreeData =
          MainEditorAssetTreeTool.BuildAssetTree.buildEmptyAssetTree();
        let addedNodeId = MainEditorAssetIdTool.getNewAssetId();

        MainEditorAssetHeaderOperateNodeTool.addScriptAttribute();

        ScriptAttributeInspectorTool.addDefaultField(
          ~sandbox,
          ~nodeId=addedNodeId,
          (),
        );

        let (fieldName, _) =
          ScriptAttributeInspectorTool.getAttributeEntries(addedNodeId)
          |> StateLogicService.getEditorState
          |> ArrayService.unsafeGetFirst;

        let newFieldName = "aaa";

        ScriptAttributeInspectorTool.renameField(
          ~sandbox,
          ~nodeId=addedNodeId,
          ~oldName=fieldName,
          ~newName=newFieldName,
          (),
        );

        (addedNodeId, (fieldName, newFieldName));
      };

      describe("test undo operate", () =>
        describe("test undo one step", () =>
          test("step which from first to zero", () => {
            let (addedNodeId, (oldFieldName, newFieldName)) = _simulate();

            RedoUndoTool.undoHistoryState();

            BuildComponentTool.buildScriptAttributeInspectorComponent(
              ~currentNodeId=addedNodeId,
              (),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          })
        )
      );

      describe("test redo operate", () =>
        describe("test redo one step", () =>
          test(
            "undo step which from first to zero, redo step which from zero to first",
            () => {
            let (addedNodeId, (oldFieldName, newFieldName)) = _simulate();

            RedoUndoTool.undoHistoryState();
            RedoUndoTool.redoHistoryState();

            BuildComponentTool.buildScriptAttributeInspectorComponent(
              ~currentNodeId=addedNodeId,
              (),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          })
        )
      );
    });

    describe("test set field data", () => {
      let _simulate = () => {
        let assetTreeData =
          MainEditorAssetTreeTool.BuildAssetTree.buildEmptyAssetTree();
        let addedNodeId = MainEditorAssetIdTool.getNewAssetId();

        MainEditorAssetHeaderOperateNodeTool.addScriptAttribute();

        ScriptAttributeInspectorTool.addDefaultField(
          ~sandbox,
          ~nodeId=addedNodeId,
          (),
        );

        let (fieldName, field) =
          ScriptAttributeInspectorTool.getAttributeEntries(addedNodeId)
          |> StateLogicService.getEditorState
          |> ArrayService.unsafeGetFirst;
        ScriptAttributeInspectorTool.updateScriptAttributeNodeByReplaceFieldData(
          addedNodeId,
          (
            fieldName,
            ScriptAttributeInspectorTool.buildFieldJsObj(
              ~type_="float",
              ~defaultValue=0.1,
            ),
          ),
        );

        (addedNodeId, (fieldName, field));
      };

      describe("test undo operate", () =>
        describe("test undo one step", () =>
          test("step which from first to zero", () => {
            let (addedNodeId, (fieldName, oldField)) = _simulate();

            RedoUndoTool.undoHistoryState();

            ScriptAttributeInspectorTool.getAttributeEntries(addedNodeId)
            |> StateLogicService.getEditorState
            |> ArrayService.unsafeGetFirst
            |> expect == (fieldName, oldField);
          })
        )
      );

      describe("test redo operate", () =>
        describe("test redo one step", () =>
          test(
            "undo step which from first to zero, redo step which from zero to first",
            () => {
            let (addedNodeId, (oldFieldName, newFieldName)) = _simulate();

            RedoUndoTool.undoHistoryState();
            RedoUndoTool.redoHistoryState();

            ScriptAttributeInspectorTool.getAttributeEntries(addedNodeId)
            |> StateLogicService.getEditorState
            |> ArrayService.unsafeGetFirst
            |> expect
            == (
                 ScriptAttributeNodeNameAssetService.getNewFieldName(),
                 ScriptAttributeInspectorTool.buildField(
                   ~type_=Wonderjs.ScriptAttributeType.Float,
                   ~defaultValue=0.1,
                 ),
               );
          })
        )
      );
    });

    describe("test remove field", () => {
      let _simulate = () => {
        let assetTreeData =
          MainEditorAssetTreeTool.BuildAssetTree.buildEmptyAssetTree();
        let addedNodeId = MainEditorAssetIdTool.getNewAssetId();

        MainEditorAssetHeaderOperateNodeTool.addScriptAttribute();

        ScriptAttributeInspectorTool.addDefaultField(
          ~sandbox,
          ~nodeId=addedNodeId,
          (),
        );

        let (fieldName, field) =
          ScriptAttributeInspectorTool.getAttributeEntries(addedNodeId)
          |> StateLogicService.getEditorState
          |> ArrayService.unsafeGetFirst;

        ScriptAttributeInspectorTool.updateScriptAttributeNodeByRemoveFieldData(
          ~sandbox,
          ~nodeId=addedNodeId,
          ~fieldName,
          (),
        );

        addedNodeId;
      };

      describe("test undo operate", () =>
        describe("test undo one step", () =>
          test("step which from first to zero", () => {
            let addedNodeId = _simulate();

            RedoUndoTool.undoHistoryState();

            ScriptAttributeInspectorTool.getAttributeEntries(addedNodeId)
            |> StateLogicService.getEngineStateToGetData
            |> Js.Array.length
            |> expect > 0;
          })
        )
      );

      describe("test redo operate", () =>
        describe("test redo one step", () =>
          test(
            "undo step which from first to zero, redo step which from zero to first",
            () => {
            let addedNodeId = _simulate();

            RedoUndoTool.undoHistoryState();
            RedoUndoTool.redoHistoryState();

            ScriptAttributeInspectorTool.getAttributeEntries(addedNodeId)
            |> StateLogicService.getEngineStateToGetData
            |> Js.Array.length
            |> expect == 0;
          })
        )
      );
    });
  });