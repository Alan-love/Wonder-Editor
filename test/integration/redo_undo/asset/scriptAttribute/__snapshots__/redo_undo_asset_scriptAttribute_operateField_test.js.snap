// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redo_undo: asset script attribute->operate field test rename field test redo operate test redo one step undo step which from first to zero, redo step which from zero to first 1`] = `
<article
  className="wonder-scriptAttribute-inspector"
>
  <h1>
    ScriptAttribute
  </h1>
  <hr />
  <article
    className="inspector-item"
  >
    <div
      className="item-header"
      title="script attribute's name"
    >
      <span
        className="component-label"
      >
        Attribute Name
      </span>
    </div>
    <div
      className="item-content"
    >
      <input
        className="input-component float-input"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="New Script Attribute"
      />
    </div>
  </article>
  <div
    className="scriptAttribute-data"
  >
    <div
      className="fields"
    >
      <div
        className="field"
      >
        <article
          className="inspector-item"
        >
          <div
            className="item-header"
            title="script attribute's field's name"
          >
            <span
              className="component-label"
            >
              Field Name
            </span>
          </div>
          <div
            className="item-content"
          >
            <input
              className="input-component float-input"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value="aaa"
            />
          </div>
        </article>
        <article
          className="wonder-file-input"
        >
          <button
            onClick={[Function]}
          >
            Set Field Data
          </button>
        </article>
        <button
          className="scriptAttribute-field-remove"
          onClick={[Function]}
        >
          Remove
        </button>
      </div>
    </div>
    <button
      className="addable-btn"
      onClick={[Function]}
    >
      Add script attribute's field
    </button>
  </div>
</article>
`;

exports[`redo_undo: asset script attribute->operate field test rename field test undo operate test undo one step step which from first to zero 1`] = `
<article
  className="wonder-scriptAttribute-inspector"
>
  <h1>
    ScriptAttribute
  </h1>
  <hr />
  <article
    className="inspector-item"
  >
    <div
      className="item-header"
      title="script attribute's name"
    >
      <span
        className="component-label"
      >
        Attribute Name
      </span>
    </div>
    <div
      className="item-content"
    >
      <input
        className="input-component float-input"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="New Script Attribute"
      />
    </div>
  </article>
  <div
    className="scriptAttribute-data"
  >
    <div
      className="fields"
    >
      <div
        className="field"
      >
        <article
          className="inspector-item"
        >
          <div
            className="item-header"
            title="script attribute's field's name"
          >
            <span
              className="component-label"
            >
              Field Name
            </span>
          </div>
          <div
            className="item-content"
          >
            <input
              className="input-component float-input"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value="New Script Attribute Field"
            />
          </div>
        </article>
        <article
          className="wonder-file-input"
        >
          <button
            onClick={[Function]}
          >
            Set Field Data
          </button>
        </article>
        <button
          className="scriptAttribute-field-remove"
          onClick={[Function]}
        >
          Remove
        </button>
      </div>
    </div>
    <button
      className="addable-btn"
      onClick={[Function]}
    >
      Add script attribute's field
    </button>
  </div>
</article>
`;
